core/trino-main/src/main/java/io/trino/operator/Driver.java

```diff
@@ -183,15 +189,15 @@ public boolean isFinished()
 
         // if we can get the lock, attempt a clean shutdown; otherwise someone else will shutdown
         Optional<Boolean> result = tryWithLockUninterruptibly(this::isFinishedInternal);
-        return result.orElseGet(() -> state.get() != State.ALIVE || driverContext.isDone());
+        return result.orElseGet(() -> state.get() != State.ALIVE || driverContext.isTerminatingOrDone());
     }
 
     @GuardedBy("exclusiveLock")
     private boolean isFinishedInternal()
     {
         checkLockHeld("Lock must be held to call isFinishedInternal");
 
-        boolean finished = state.get() != State.ALIVE || driverContext.isDone() || activeOperators.isEmpty() || activeOperators.get(activeOperators.size() - 1).isFinished();
+        boolean finished = state.get() != State.ALIVE || activeOperators.isEmpty() || activeOperators.get(activeOperators.size() - 1).isFinished() || driverContext.isTerminatingOrDone();
```

Comment:
I'm ok with that, but the notion of "terminatingOr(Done,Finished)" doesn't apply super cleanly through all the usage sites. `Driver` has one specific definition (that's pretty complex in terms of what it entails), and `SplitRunner` has another that mostly overlaps, but also includes whether `DriverSplitRunner#close()` was called, and finally `PrioritizedSplitRunner` has another notion that includes whether `PrioritizedSplitRunner#destroy()` was called- so ultimately I think it's probably fine to keep the notion of `terminatingOrDone()` (task level concern) separate from `isFinished()` (`Driver`, `SplitRunner`, `PrioritizedSplitRunner` concern).

Reply:
Yeah, I can see that. What do you think about changing it for the `Driver` only?

Reply:
I think maybe the compromise is to rename `isFinishedInternal()` but not `Driver#isFinished()` since the internal method _is_ about checking whether the driver _should_ be destroyed (implies association to "terminating") but `Driver#isFinished()` is a separate semantic that needs only to describe having the Driver itself having "finished" (for whatever reason).

Reply:
Is there a reason this was reordered?  I'm just curious if the order matters here.  (if it does add a comment)

Reply:
> Is there a reason this was reordered? I'm just curious if the order matters here.

No correctness reason.

I reordered these checks originally while trying to disentangle race conditions and the isTerminatingOrDone() check used to eventually synchronize on TaskHandle.

At this point that check is now just a volatile read (landed in another PR, works similarly to how DriverFactory does in this PR), but I chose to keep the new ordering because it still holds that “normal” finishing is more likely than task termination and so the cheaper checks that of driver-local state should come first before trying to check shared state (or acquire shared locks).